engine/src/flutter/display_list/geometry/dl_path.cc

```diff
@@ -252,26 +291,94 @@ DlPath DlPath::operator+(const DlPath& other) const {
+  // degenerate conics gracefully (or have all receivers just handle
+  // them directly). But, until then, we will just convert them to a
+  // pair of quads and accept the results as "close enough".
+  if (p1 != cp) {
+    if (cp != p2) {
+      std::array<DlPoint, 5> points;
+      impeller::ConicPathComponent conic(p1, cp, p2, weight);
+      conic.SubdivideToQuadraticPoints(points);
+      receiver.QuadTo(points[1], points[2]);
+      receiver.QuadTo(points[3], points[4]);
+    } else {
+      receiver.LineTo(cp);
+    }
+  } else if (cp != p2) {
+    receiver.LineTo(p2);
+  }
+}
+
+namespace {
+class SkiaPathReceiver final : public DlPathReceiver {
+ public:
+  void SetPathInfo(DlPathFillType fill_type, bool is_convex) override {
+    sk_path_.setFillType(ToSkFillType(fill_type));
+  }
+  void MoveTo(const DlPoint& p2) override { sk_path_.moveTo(ToSkPoint(p2)); }
+  void LineTo(const DlPoint& p2) override { sk_path_.lineTo(ToSkPoint(p2)); }
+  void QuadTo(const DlPoint& cp, const DlPoint& p2) override {
+    sk_path_.quadTo(ToSkPoint(cp), ToSkPoint(p2));
+  }
+  bool ConicTo(const DlPoint& cp, const DlPoint& p2, DlScalar weight) override {
+    sk_path_.conicTo(ToSkPoint(cp), ToSkPoint(p2), weight);
+    return true;
+  }
+  void CubicTo(const DlPoint& cp1,
+               const DlPoint& cp2,
+               const DlPoint& p2) override {
+    sk_path_.cubicTo(ToSkPoint(cp1), ToSkPoint(cp2), ToSkPoint(p2));
+  }
+  void Close() override { sk_path_.close(); }
+
+  SkPath TakePath() { return sk_path_; }
+
+ private:
+  SkPath sk_path_;
+};
+}  // namespace
+
+SkPath DlPath::ConvertToSkiaPath(const Path& path) {
+  SkiaPathReceiver receiver;
+
+  DispatchFromImpellerPath(path, receiver);
+
+  return receiver.TakePath();
+}
+
+void DlPath::DispatchFromImpellerPath(const impeller::Path& path,
+                                      DlPathReceiver& receiver) {
   bool subpath_needs_close = false;
   std::optional<DlPoint> pending_moveto;
 
-  auto resolve_moveto = [&pending_moveto, &sk_path]() {
+  auto resolve_moveto = [&receiver, &pending_moveto]() {
     if (pending_moveto.has_value()) {
-      sk_path.moveTo(ToSkPoint(pending_moveto.value()));
+      receiver.MoveTo(pending_moveto.value());
       pending_moveto.reset();
     }
   };
 
+  // The Impeller Point Count is way overestimated due to duplicate
+  // points between elements.
+  receiver.RecommendSizes(path.GetComponentCount(), path.GetPointCount());
+  std::optional<DlRect> bounds = path.GetBoundingBox();
+  if (bounds.has_value()) {
+    receiver.RecommendBounds(bounds.value());
+  }
+  receiver.SetPathInfo(path.GetFillType(), path.IsConvex());
   for (auto it = path.begin(), end = path.end(); it != end; ++it) {
     switch (it.type()) {
       case ComponentType::kContour: {
         const impeller::ContourComponent* contour = it.contour();
         FML_DCHECK(contour != nullptr);
         if (subpath_needs_close) {
-          sk_path.close();
+          receiver.Close();
```

Comment:
No action required:  It's not in the c++ style guide but when calling virtual methods it's safer to use `->` since if you accidentally get a value instead of a reference you'll get incorrect behavior.

Reply:
How? `->` doesn't even compile.

Reply:
You have to switch from using references to using pointers.

Reply:
That would add the burden of implicitly accepting null values which must be checked. I don't understand the "value vs reference" issue and the incorrect behavior it leads to.

Reply:
When you have a reference in C++ it isn't guaranteed to not be null.  You still technically have to consider null values, we just aren't ignoring that.

Reply:
Perhaps clang-tidy might complain (if it is run)?

Reply:
> but I'm guessing you can fudge it with *pointer_that_is_null?

Yep, exactly and we do this in our codebase all the time.  A linter that disallowed references to anything on the heap would be nice, I don't think it exists.  That's how I personally use them.

Reply:
I'll file this under "wishing C++ would throw if you create a reference by dereferencing a nulllptr".